libkbfs/folder_branch_ops.go

```diff
@@ -4074,53 +4069,92 @@ func (fbo *folderBranchOps) searchForNode(ctx context.Context,
 	return n, nil
 }
 
-func (fbo *folderBranchOps) unlinkFromCache(op op, unlinkPath path) error {
-	// The entry could be under any one of the unref'd blocks, and
-	// it's safe to perform this when the pointer isn't real, so just
-	// try them all to avoid the overhead of looking up the right
-	// pointer in the old version of the block.
-	for _, ptr := range op.Unrefs() {
-		// It's ok to modify this path, since we break as soon as the
-		// node cache takes a reference to it.
-		unlinkPath.path[len(unlinkPath.path)-1].BlockPointer = ptr
-		found := fbo.nodeCache.Unlink(ptr.Ref(), unlinkPath)
-		if found {
-			break
-		}
-	}
+func (fbo *folderBranchOps) unlinkFromCache(
+	op op, unlinkPath path, unlinkDe DirEntry) error {
 
 	return nil
 }
 
 func (fbo *folderBranchOps) getUnlinkPathBeforeUpdatingPointers(
-	ctx context.Context, op op) (unlinkPath path, toUnlink bool, err error) {
+	ctx context.Context, lState *lockState, md ReadOnlyRootMetadata, op op) (
+	unlinkPath path, unlinkDe DirEntry, toUnlink bool, err error) {
 	var node Node
 	var childName string
 
+	requireResFix := false
 	switch realOp := op.(type) {
 	case *rmOp:
+		if realOp.Dir.Ref == realOp.Dir.Unref {
+			requireResFix = true
+		}
 		node = fbo.nodeCache.Get(realOp.Dir.Unref.Ref())
 		childName = realOp.OldName
 	case *renameOp:
 		if realOp.NewDir.Unref != zeroPtr {
 			// moving to a new dir
+			if realOp.NewDir.Ref == realOp.NewDir.Unref {
+				requireResFix = true
+			}
 			node = fbo.nodeCache.Get(realOp.NewDir.Unref.Ref())
 		} else {
 			// moving to the same dir
+			if realOp.OldDir.Ref == realOp.OldDir.Unref {
+				requireResFix = true
+			}
 			node = fbo.nodeCache.Get(realOp.OldDir.Unref.Ref())
 		}
 		childName = realOp.NewName
 	}
 	if node == nil {
-		return path{}, false, nil
+		return path{}, DirEntry{}, false, nil
 	}
 
 	p, err := fbo.pathFromNodeForRead(node)
 	if err != nil {
-		return path{}, false, err
+		return path{}, DirEntry{}, false, err
 	}
 
-	return p.ChildPathNoPtr(childName), true, nil
+	// If the first op in this MD update is a resolutionOp, we need to
+	// inspect it to look for the *real* original pointer for this
+	// node.
+	if resOp, ok := md.data.Changes.Ops[0].(*resolutionOp); ok {
```

Comment:
Yeah, we can assume that `op` came from `md`, so we know at least there's one.  But I can add a defensive check to make sure.